src/Uno.UI/UI/Xaml/UIElement.wasm.cs

```diff
@@ -622,8 +622,8 @@ internal void MoveChildTo(int oldIndex, int newIndex)
 
 			RegisterEventHandler(
 				domEventName,
-				handler: new RoutedEventHandlerWithHandled((snd, args) => RaiseEvent(routedEvent, args)),
-				invoker: GenericEventHandlers.RaiseRoutedEventHandlerWithHandled,
+				handler: new RoutedEventHandler((snd, args) => RaiseEvent(routedEvent, args)),
+				invoker: GenericEventHandlers.RaiseRoutedEventHandler,
```

Comment:
This is not valid: you are only ignoring the [`Handled`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.input.keyroutedeventargs.handled?view=winrt-20348) flag here. It means that a key event flag as handled would still bubble ...

Reply:
@dr1rrb Thanks! However, I don't feel I'm fully understanding the situation here (sorry). Is the suggestion is to modify this:

https://github.com/unoplatform/uno/blob/bc8d7fa884184c0aad3af1e4763d1f929757d503/src/Uno.UI/UI/Xaml/UIElement.GenericEventHandlers.wasm.cs#L56-L57

to check for `RoutedEventArgs.Handled` property, then decide whether to `return null` or `return handler(....)`?

Or get rid of `RaiseRoutedEventHandlerWithHandled` and use only `RaiseRoutedEventHandler` which then should behave differently based on `RoutedEventArgs.Handled` (ie, decide whether to `return null` or `return handler(....)`)?

Or something else?

Reply:
@Youssef1313 There is something that is really important about "routed" events, is that they are _bubbling_ from the top-most element (a.k.a. the leaf of the visual tree) up to the root of the visual tree. It means that for instance when you press the button of your mouse, the element right under the cursor will fire the `PointerPressed` event, then its parent will do the same, then its parent etc. until we reach the root of the visual tree. That's the case for almost all input events (pointers, keyboard), and a lot of so called "routed events" like `GotFocus` (with a noticeable exception of the [`PreviewKeyDown` event](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.previewkeydown?view=winrt-20348) which is the single _tunneling_ event on UWP, which means that the event is raised on the root first, then on children until we reach the leaf of the tree).

The args of those "routed" event have a `Handled` flag that allow any handler to stop that _bubbling_ at any time. This is really useful when, for instance, you have a button in a button: the "top most" button will handle the pointer events, so the `Click` will be fired only on one button.

In browsers we also have a second layer of "handling":
* `stopPropagation` is really the same thing as flagging an event `Handled` on UWP: it just stops the _bubbling_
* `preventDefault` which request to the browser to not apply the default behavior that it would normally do (scrolling, navigation, focus, etc.). 

Be aware that "stopping propagation" will **NOT** prevent the default behavior. Both flags are not linked in any way.
And the difference is really important. A good example is for the touch pointers: if an application decides to handle a `PointerMoved` event, we must not prevent the default behavior since for touch it will be to scroll the content (if possible).

Unfortunately we don't have a such distinction in UWP API (which is normal since there is no "default behavior" like that on UWP), so publicly we are mapping both to the single UWP `Handled` flag, which is enough so far. That's why we have [`HtmlEventDispatchResult.StopPropagation | HtmlEventDispatchResult.PreventDefault` here](https://github.com/unoplatform/uno/blob/9efab96fd59e28c68fa1430f7cbc74353ebc22bf/src/Uno.UI/UI/Xaml/UIElement.EventRegistration.wasm.cs#L155). And internally we are able to specify the `HtmlEventDispatchResult`, but only using the `RawEventHandler` for now.

In the https://github.com/unoplatform/uno/blob/bc8d7fa884184c0aad3af1e4763d1f929757d503/src/Uno.UI/UI/Xaml/UIElement.GenericEventHandlers.wasm.cs basically we just have an enumeration of supported strongly typed event handlers. Each handler was designed for some specific usage:
* The `EventHandler` is for non routed events
* The `RawEventHandler` is the more powerful since you are able to get the raw arguments without parsing, and explicitly specify the `HtmlEventDispatchResult`.
* The `RoutedEventHandler` is for routed event args that does not have the `Handled` flag (which should be really rare ... maybe it's  even for legacy support)
* The `ExceptionRoutedEventHandler` don't even know why it exists neither why it's named "routed"
* The `RoutedEventHandlerWithHandled` is the common one for routed events: it just returns a `bool` which is actually the `Handled` flag which is then converted in JS as "stop bubbling" AND "prevent default behavior" as describe above.

The main "fix" is not really in those files: It's to find which handler is flagging the key events as `Handled` and change that in order to make a distinction between "stop propagation" and "prevent default", but **only for that specific event handler**. If it's in the focus management like I suppose, I would assume that we just don't want to "prevent default" when the focused element is a `TextBox`.

Reply:
@dr1rrb Thank you a LOT for the detailed explanation! Really much appreciated.

If I got this right, the problem is in:

https://github.com/unoplatform/uno/blob/0789a59d33760ba6997640deeb50c03e20d1e676/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs#L665-L688

However, the behavior seems like it was intentional, so I assume it's not as simple as removing the override. Any idea why this code used to exist?

Reply:
Looks like this was aligning the behavior with UWP: https://github.com/unoplatform/uno/pull/684

Reply:
Ok then from what I see here, we want to flag the events as "handled" to avoid bubbling in managed code (like in UWP does apparently). And in the PR we also made sure to **not** `preventDefault` when event was flagged as handled ... which is obviously no longer the case. So it means that in that event handler of the `TextBox` we should have a way to specify that on wasm, we don't want to `preventDefault`.

The easiest way would be to add the ability to set the `HtmlEventDispatchResult` on the [internal interface `IHandleableRoutedEventArgs`](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Input/IHandleableRoutedEventArgs.cs) (for `__WASM__` only) and then make sure to send the result back in event handlers to reach [that case](https://github.com/unoplatform/uno/blob/9efab96fd59e28c68fa1430f7cbc74353ebc22bf/src/Uno.UI/UI/Xaml/UIElement.EventRegistration.wasm.cs#L157). Setting `IHandleableRoutedEventArgs.Handled = true` would then just set `HtmlEventDispatchResult.StopPropagation | HtmlEventDispatchResult.PreventDefault` ... the issue is that there are a lot of classes which are implementing that interface, maybe the default interface implementation might help you here.

Reply:
@dr1rrb Thanks again! I'll take a look and see what I can do.

btw, the code above isn't as what you originally wrote. `AcceptsReturn` check didn't use to exist. I assume this could be a bug introduced later? Can you confirm?

Also, is there anyway to debug on Wasm? I was never able to get that working. `Console.WriteLine`s going to be very very tedious here.

Reply:
I wasn't able to add the ability to set `HtmlEventDispatchResult`. The problem I faced is that `RoutedEventHandlerWithHandled` delegate returns a `bool`. I tried to take a different (but relatively similar) approach. But it doesn't seem to work. Trying to figure that out.

Reply:
> 
> Also, is there anyway to debug on Wasm? I was never able to get that working. `Console.WriteLine`s going to be very very tedious here.

You should be able just to press `F5` : https://platform.uno/blog/debugging-uno-platform-webassembly-apps-in-visual-studio-2019/ ... but the debugger is still not so nice and you will most probably end have to add some `Console.WriteLine` at some point.

Reply:
F5 doesn't seem to work for me.

![image](https://user-images.githubusercontent.com/31348972/134119095-c0af61e6-8f79-4759-a64b-b28f9cd5c171.png)

Reply:
The debugger is not updated yet in the sample app. Try testing your PR in a separate app, with the Uno.Wasm.Bootstrap package updated to 3.0.0.